File: frontend/src/lib/budget.ts
export type PeriodType = 'week' | 'month';
export type BudgetKind = 'fixed' | 'flexible' | 'discretionary';
export interface BudgetItem {
  id: string; categoryId: string; limitCents: number; warningRatio: number;
  kind: BudgetKind; note: string | null;
}
export interface BudgetPlan {
  id: string; periodType: PeriodType; startsOn: string; endsOn: string; currency: string;
  plannedIncomeCents: number | null; savingsTargetCents: number | null;
  status: 'draft' | 'active' | 'closed'; note: string | null; items: BudgetItem[];
}
export interface BudgetSummary {
  planId: string; currency: string; timezone: string;
  period: { type: PeriodType; startsOn: string; endsOn: string; daysElapsed: number; daysRemaining: number };
  plannedIncomeCents: number | null; savingsTargetCents: number | null;
  plannedSpendCents: number; plannedResidualCents: number | null; actualSpendCents: number;
  remainingBudgetCents: number; projectedSpendCents: number; projectedSavingsCents: number | null;
  safeDailySpendCents: number;
  items: { categoryId: string; kind: BudgetKind; limitCents: number; spentCents: number;
    remainingCents: number; usedRatio: number | null; paceRatio: number | null;
    status: 'safe' | 'watch' | 'warning' | 'over' }[];
  unbudgeted: { categoryId: string; spentCents: number }[];
}

export function todayInZone(timezone = 'Asia/Tokyo'): string {
  const parts = new Intl.DateTimeFormat('en-US', { timeZone: timezone, year: 'numeric', month: '2-digit', day: '2-digit' }).formatToParts(new Date());
  const value = (type: string) => parts.find(p => p.type === type)?.value;
  return `${value('year')}-${value('month')}-${value('day')}`;
}

// Date-only navigation uses UTC to avoid browser timezone and DST drift.
export function periodBounds(type: PeriodType, on: string): { start: string; end: string } {
  const d = new Date(`${on}T00:00:00Z`);
  if (type === 'week') d.setUTCDate(d.getUTCDate() - (d.getUTCDay() + 6) % 7);
  else d.setUTCDate(1);
  const start = d.toISOString().slice(0, 10);
  if (type === 'week') d.setUTCDate(d.getUTCDate() + 6);
  else d.setUTCMonth(d.getUTCMonth() + 1, 0);
  return { start, end: d.toISOString().slice(0, 10) };
}

export function shiftPeriod(type: PeriodType, on: string, delta: number): string {
  const d = new Date(`${periodBounds(type, on).start}T00:00:00Z`);
  if (type === 'week') d.setUTCDate(d.getUTCDate() + delta * 7);
  else d.setUTCMonth(d.getUTCMonth() + delta);
  return d.toISOString().slice(0, 10);
}

/** Parse decimal input exactly; money is stored as major units ×100 for every currency. */
export function parseMoney(value: string): number {
  if (!/^\d+(\.\d{1,2})?$/.test(value.trim())) throw new Error('Enter a non-negative amount with at most two decimal places.');
  const [major, minor = ''] = value.trim().split('.');
  const result = Number(major) * 100 + Number(minor.padEnd(2, '0'));
  if (!Number.isSafeInteger(result)) throw new Error('Amount is too large.');
  return result;
}

File: backend/api/schemas/budget.py
from datetime import date, datetime
from typing import Literal

from pydantic import ConfigDict, Field, StrictInt

from backend.api.schemas.base import CamelModel

PeriodType = Literal["week", "month"]
Kind = Literal["fixed", "flexible", "discretionary"]


class BudgetInput(CamelModel):
    model_config = ConfigDict(extra="forbid")


class PlanCreate(BudgetInput):
    period_type: PeriodType
    starts_on: date
    ends_on: date | None = None
    currency: str
    planned_income_cents: StrictInt | None = None
    savings_target_cents: StrictInt | None = None
    status: Literal["draft", "active", "closed"] = "active"
    note: str | None = None


class PlanUpdate(BudgetInput):
    planned_income_cents: StrictInt | None = None
    savings_target_cents: StrictInt | None = None
    status: Literal["draft", "active", "closed"] | None = None
    note: str | None = None


class ItemCreate(BudgetInput):
    category_id: str
    limit_cents: StrictInt
    warning_ratio: float = Field(default=0.8, gt=0, le=1)
    kind: Kind = "flexible"
    note: str | None = None


class ItemUpdate(BudgetInput):
    limit_cents: StrictInt | None = None
    warning_ratio: float | None = Field(default=None, gt=0, le=1)
    kind: Kind | None = None
    note: str | None = None


class ItemOut(ItemCreate):
    id: str
    plan_id: str
    created_at: datetime
    updated_at: datetime


class PlanOut(PlanCreate):
    id: str
    ends_on: date
    created_by: str
    created_at: datetime
    updated_at: datetime
    items: list[ItemOut]


class CloneInput(BudgetInput):
    starts_on: date
    ends_on: date | None = None


class QuickInput(BudgetInput):
    amount_cents: StrictInt
    category_id: str
    currency: str
    occurred_on: date
    note: str | None = None


class SetTotalInput(BudgetInput):
    plan_id: str
    category_id: str
    total_cents: StrictInt
    note: str | None = None


class PeriodOut(CamelModel):
    type: PeriodType
    starts_on: date
    ends_on: date
    days_elapsed: int
    days_remaining: int


class ItemSummary(CamelModel):
    category_id: str
    kind: Kind
    limit_cents: int
    spent_cents: int
    remaining_cents: int
    used_ratio: float | None
    pace_ratio: float | None
    status: Literal["safe", "watch", "warning", "over"]


class Unbudgeted(CamelModel):
    category_id: str
    spent_cents: int


class SummaryOut(CamelModel):
    plan_id: str
    currency: str
    timezone: str
    period: PeriodOut
    planned_income_cents: int | None
    savings_target_cents: int | None
    planned_spend_cents: int
    planned_residual_cents: int | None
    actual_spend_cents: int
    remaining_budget_cents: int
    projected_spend_cents: int
    projected_savings_cents: int | None
    safe_daily_spend_cents: int
    items: list[ItemSummary]
    unbudgeted: list[Unbudgeted]
